File: android_java_hummus/src/main/java/com/github/klee0kai/hummus/adapterdelegates/delegate/ClassAdapterDelegate.java
package com.github.klee0kai.hummus.adapterdelegates.delegate;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;


import com.github.klee0kai.hummus.adapterdelegates.SimpleViewHolder;
import com.hannesdorfmann.adapterdelegates4.AdapterDelegate;

import java.util.List;

public abstract class ClassAdapterDelegate<T> extends AdapterDelegate<List<Object>> {

    @LayoutRes
    private final int layoutRes;
    private final Class<T> tClass;

    public ClassAdapterDelegate(Class<T> tClass, @LayoutRes int layoutRes) {
        this.tClass = tClass;
        this.layoutRes = layoutRes;
    }

    @Override
    protected boolean isForViewType(@NonNull List<Object> items, int position) {
        return tClass.isInstance(items.get(position));
    }

    @NonNull
    @Override
    protected SimpleViewHolder onCreateViewHolder(ViewGroup parent) {
        return new SimpleViewHolder(LayoutInflater.from(parent.getContext()).inflate(layoutRes, parent, false));
    }

    @Override
    protected void onBindViewHolder(@NonNull List<Object> objects, int i, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull List<Object> list) {
        onBindViewHolder((T) objects.get(i), i, (SimpleViewHolder) viewHolder, objects, list);
    }


    protected abstract void onBindViewHolder(T it, int pos, @NonNull SimpleViewHolder vh, @NonNull List<Object> allObjects, @NonNull List<Object> allOldlist);

}

File: android_kotlin_hummus/src/main/kotlin/com/github/klee0kai/hummus/adapterdelegates/delegate/ClassAdapterDelegate.kt
package com.github.klee0kai.hummus.adapterdelegates.delegate

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.github.klee0kai.hummus.adapterdelegates.SimpleViewHolder
import com.hannesdorfmann.adapterdelegates4.AdapterDelegate


abstract class ClassAdapterDelegate<T>(
    protected open val tClass: Class<T>,
    @field:LayoutRes @param:LayoutRes protected open val layoutRes: Int,
    protected open val forceTheSameClass: Boolean = false
) : AdapterDelegate<List<Any>>() {


    override fun isForViewType(items: List<Any>, position: Int): Boolean {
        val item = items[position]
        return if (forceTheSameClass) {
            tClass == item.javaClass
        } else {
            tClass.isInstance(item)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup): SimpleViewHolder {
        return SimpleViewHolder(
            LayoutInflater.from(parent.context).inflate(
                layoutRes, parent, false
            )
        )
    }

    override fun onBindViewHolder(
        objects: List<Any>,
        pos: Int,
        viewHolder: ViewHolder,
        list: List<Any>
    ) {
        onBindViewHolder(objects[pos] as T, pos, viewHolder as SimpleViewHolder)
    }

    protected abstract fun onBindViewHolder(it: T, pos: Int, vh: SimpleViewHolder)
}